pkg/hubble/exporter/cell.go

```diff
@@ -0,0 +1,161 @@
+	flags.StringSlice("hubble-export-fieldmask", def.ExportFieldmask, "Specify list of fields to use for field mask in Hubble exporter.")
+}
+
+func (cfg config) validate() error {
+	if fm := cfg.ExportFieldmask; len(fm) > 0 {
+		_, err := fieldmaskpb.New(&flowpb.Flow{}, fm...)
+		if err != nil {
+			return fmt.Errorf("hubble-export-fieldmask contains invalid fieldmask '%v': %w", fm, err)
+		}
+	}
+	return nil
+}
+
+type hubbleExportersOut struct {
+	cell.Out
+
+	// use flatten as a poor man's optional field.
+	ObserverOptions []observeroption.Option `group:"hubble-observer-options,flatten"`
+}
+
+type hubbleExportersParams struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	Config    config
+
+	OnExportEvent []exporteroption.OnExportEvent `group:"hubble-exporter-on-export-event"`
+
+	// TODO: replace by slog
+	Logger logrus.FieldLogger
+}
+
+func newHubbleStaticExporter(params hubbleExportersParams) (hubbleExportersOut, error) {
+	if err := params.Config.validate(); err != nil {
+		return hubbleExportersOut{}, fmt.Errorf("hubble-exporter configuration error: %w", err)
+	}
+
+	if params.Config.ExportFilePath != "" {
+		exporterOpts := []exporteroption.Option{
+			exporteroption.WithPath(params.Config.ExportFilePath),
+			exporteroption.WithMaxSizeMB(params.Config.ExportFileMaxSizeMB),
+			exporteroption.WithMaxBackups(params.Config.ExportFileMaxBackups),
+			exporteroption.WithAllowList(params.Logger, params.Config.ExportAllowlist),
+			exporteroption.WithDenyList(params.Logger, params.Config.ExportDenylist),
+			exporteroption.WithFieldMask(params.Config.ExportFieldmask),
+		}
+		if params.Config.ExportFileCompress {
+			exporterOpts = append(exporterOpts, exporteroption.WithCompress())
+		}
+		for _, onExportEvent := range params.OnExportEvent {
+			exporterOpts = append(exporterOpts, exporteroption.WithOnExportEvent(onExportEvent))
+		}
+		exporter, err := NewExporter(params.Logger, exporterOpts...)
+		if err == nil {
+			params.Lifecycle.Append(cell.Hook{
+				OnStop: func(hc cell.HookContext) error {
+					return exporter.Stop()
+				},
+			})
+			return hubbleExportersOut{
+				ObserverOptions: []observeroption.Option{observeroption.WithOnDecodedEvent(exporter)},
+			}, nil
+		}
+		params.Logger.WithError(err).Error("Failed to configure Hubble static exporter")
+	}
+
+	return hubbleExportersOut{}, nil
+}
+
+func newHubbleDynamicExporter(params hubbleExportersParams) (hubbleExportersOut, error) {
+	if err := params.Config.validate(); err != nil {
+		return hubbleExportersOut{}, fmt.Errorf("hubble-exporter configuration error: %w", err)
+	}
+
+	if params.Config.FlowlogsConfigFilePath != "" {
+		exporter := NewDynamicExporter(params.Logger, params.Config.FlowlogsConfigFilePath, params.Config.ExportFileMaxSizeMB, params.Config.ExportFileMaxBackups)
+		params.Lifecycle.Append(cell.Hook{
+			OnStart: func(hc cell.HookContext) error {
+				go exporter.Watch()
```

Comment:
I was actually going to ask, because it seems other hive cells are mixed between doing this and using job.OneShot. I assume OneShot is better for long-lived task whereas OnStart/OnStop is essentially doing point-in-time work that can block ?

With the recent refactor, we can do both, so let me update it.

Reply:
Updated to use job.OneShot, much nicer thanks.